### ISSUE TITLE
[Bug] SDPA's default chunk size of 32 costs up to 3.75x in time on shipped model paths

### ISSUE THREAD

**Issue by kinginu** [tenstorrent/tt-metal#51911]
**Component / Area:** ttnn ops (transformer / SDPA prefill)

**Issue Type:** Bad Outputs and Performance

## Observed

`ttnn.transformer.scaled_dot_product_attention` called **without a `program_config`** uses `q_chunk_size = k_chunk_size = 32` — one tile, the smallest legal value. On default arguments, against unmodified `main`, the op is **up to 3.75x slower** than it is with chunk sizes derived from the shape, and its attention probabilities do not sum to 1.

The speed is the substantial half. On the accuracy, see the caveat at the end of this section — it is real but smaller than the row sums make it look, and I would not file on it alone.

Blackhole P300, `bfloat16`, head_dim 128, `torch.randn` inputs. `dev` is the worst row's distance from 1.0; "derived" means chunk sizes taken from the shape. 10 calls back to back with one synchronize, median of 3.

| b | heads | seq | causal | dev, default | derived | ms, default | derived | speedup |
|---|---|---|---|---|---|---|---|---|
| 1 | 1 | 2048 | ✗ | 0.0391 | 0.0234 | 0.126 | 0.097 | 1.30x |
| 1 | 1 | 4096 | ✗ | 0.0508 | 0.0234 | 0.434 | 0.198 | 2.19x |
| 1 | 1 | 8192 | ✗ | 0.2031 | 0.0234 | 1.283 | 0.405 | 3.17x |
| 1 | 1 | 8192 | ✓ | 0.1875 | 0.0234 | 1.648 | 0.569 | 2.90x |
| 1 | 1 | 16384 | ✗ | 0.2070 | 0.0312 | 4.256 | 1.218 | 3.49x |
| 1 | 1 | 16384 | ✓ | 0.2578 | 0.0273 | 5.518 | 1.471 | **3.75x** |
| 1 | 8 | 8192 | ✗ | 0.2188 | 0.0273 | 9.757 | 3.043 | 3.21x |
| 1 | 8 | 8192 | ✓ | 0.1875 | 0.0234 | 10.996 | 3.105 | 3.54x |

With the heavier-tailed inputs the SDPA tests already use, the worst row at sequence 16384 causal sums to **0.625** and the largest to 1.266.

**What that costs the output is much less than it sounds, and I would rather say so than have it found later.** Against an fp32 reference with real `V`, PCC moves from 0.99932 at the default to 0.99968 with the derived sizes — both comfortably inside any threshold in the tree. Per-row relative error is 5x worse at the default (0.1333 against 0.0251 at sequence 16384), but roughly three quarters of that excess is a **per-row scale**: fit the best per-row gain and remove it, and 0.1333 falls to 0.0344 while the derived sizes barely move (0.0251 to 0.0246). A per-row gain is the component a linear projection and an RMSNorm absorb most of, so what survives downstream is more like 40% extra residual than 5x.

| seq 16384 | q/k | PCC vs fp32 | err/row | after removing per-row scale | per-row gain spread |
|---|---|---|---|---|---|
| non-causal | 32/32 | 0.999324 | 0.1333 | 0.0344 | 0.1111 |
| non-causal | 128/512 | 0.999683 | 0.0257 | 0.0250 | 0.0516 |
| causal | 32/32 | 0.999320 | 0.0591 | 0.0288 | 0.2221 |
| causal | 128/512 | 0.999743 | 0.0249 | 0.0242 | 0.0575 |

One piece is not a scale: in the causal case the residual is position-dependent, 0.80 for the first half of the query positions against the second at the default, and 0.96 once the fold count comes down. So the default weights early and late context slightly differently. I have not shown that this changes anything a model produces.

The 32 dates to `1901aedf96b` (2024-08-13, *"#10136: Move SDPA ops to ttnn"*) and is unchanged since. If it is deliberate, that changes what the right fix looks like — please say so.

34 configurations measured — head dims 64/256, `bfloat8_b`, batch 2, GQA 8:1 and 8:2 — accuracy improves in every one.

Two things ruled out:

* **Compute config.** `fp32_dest_acc_en` does not help and at sequence 16384 makes it worse (row sum mean 0.9633 → 0.9108); `exp_approx_mode` moves it by 3e-5.
* **`q_chunk_size`.** 32 through 512 gives row sums identical to five decimals. The accuracy axis is `k_chunk_size` alone.

## Expected

A default that is not several times slower than an obvious alternative on the same shape, and attention probabilities that sum to 1.

A default derived from the shape, the way `sdpa_decode` already picks its chunk size (`sdpa_decode.cpp`, `get_chunk_size`), bounded by L1, the sliding window and a floor on the k-chunk count. Sequences short enough that the fold count is small should keep 32/32, so nothing changes for them.

## Steps to reproduce

Set `V` to all ones. Then `out = P @ V` is exactly the row sums of the attention probabilities, so the op's output **is** the normalisation its softmax arrived at. It must be 1.0.

```python
import torch, ttnn

d = ttnn.open_device(device_id=0)
seq, dh = 16384, 128
torch.manual_seed(0)

def fa_rand(*shape):  # the distribution tests/ttnn/unit_tests/operations/sdpa/ already uses
    return torch.randn(shape) + torch.randn(shape) * 10 * torch.bernoulli(torch.full(shape, 0.001))

q = fa_rand(1, 1, seq, dh).to(torch.bfloat16)
k = fa_rand(1, 1, seq, dh).to(torch.bfloat16)
v = torch.ones((1, 1, seq, dh), dtype=torch.bfloat16)   # out = P @ V == row sums of P
tq, tk, tv = (ttnn.from_torch(z, layout=ttnn.TILE_LAYOUT, device=d) for z in (q, k, v))

for chunk in (None, 128, 512):
    kw = {} if chunk is None else {"program_config": ttnn.SDPAProgramConfig(
        compute_with_storage_grid_size=d.compute_with_storage_grid_size(),
        q_chunk_size=128, k_chunk_size=chunk, exp_approx_mode=True)}
    o = ttnn.to_torch(ttnn.from_device(
        ttnn.transformer.scaled_dot_product_attention(tq, tk, tv, is_causal=True, **kw))).float()
    print(f"k_chunk={str(chunk):>4}: mean {float(o.mean()):.5f}  min {float(o.min()):.5f}  max {float(o.max()):.5f}")
ttnn.close_device(d)
```

Measured on Blackhole P300:

```
k_chunk=None: mean 1.02290  min 0.62500  max 1.26562
k_chunk= 128: mean 0.99529  min 0.83984  max 1.13281
k_chunk= 512: mean 0.99767  min 0.95312  max 1.06250
```

`k_chunk=None` is the default path. Nothing but the chunk size differs.

## Model level: gemma4 prefill

**Corrected after filing:** gemma4's single-user prefill is already tuned. `prefill_forward` delegates to `_prefill_forward_single` when `batch_size <= 1`, and that path passes `prefill_sdpa_program_config` at three call sites; its docstring already notes that explicit chunks cut the per-chunk overhead. What is still on the default is the **batched multi-user** path (`prefill.py:521`) and `chunked_prefill_sdpa_sliding` (`operations.py:398`).

The numbers below are therefore SDPA at gemma4's shapes and layer mix with and without a config — which is what the batched path pays today, not what a single-user prefill pays. SDPA at gemma4's per-layer shapes, summed over the layer mix, both from `models/demos/gemma4/configs/*/config.json`:

| model | layer mix | seq | SDPA per forward, default | derived | speedup |
|---|---|---|---|---|---|
| E2B (8q/1kv, dh 256, win 512) | 28 sliding + 7 full | 4096 | 81.2 ms | 29.6 ms | 2.75x |
| | | 8192 | 233.2 ms | 70.0 ms | 3.33x |
| | | 16384 | 783.5 ms | 214.8 ms | 3.65x |
| 12B (16q/8kv, dh 256, win 1024) | 40 sliding + 8 full | 16384 | 2283.6 ms | 610.6 ms | 3.74x |
| 31B (32q/16kv, dh 256, win 1024) | 50 sliding + 10 full | 8192 | 1968.0 ms | 547.8 ms | 3.59x |
| | | 16384 | **5719.0 ms** | **1565.7 ms** | 3.65x |

Full-attention layers also stop losing mass: worst deviation at 16384 goes from 0.2969–0.3203 to 0.0430–0.0469.

**Shows:** real shapes, head counts, window sizes and layer counts, on the paths still using the default. 4.15 s per forward pass at 31B / 16k.

**Does not show:** this is not a model run — the image's `transformers` does not know the `gemma4` architecture, so I called SDPA directly at the config's shapes.

* attention only, so not an end-to-end speedup;
* no PCC against a reference — the accuracy column is normalisation, not output quality;
* layers timed separately and summed, so no trace amortisation;
* single chip.

## Affected call sites

63 `ttnn.transformer.scaled_dot_product_attention` call sites under `models/`, 5 of them in `reference/` or `tests/`. Of the 58 in shipped model code, 47 pass a `program_config` and **11 do not**:

```
models/demos/gemma4/tt/attention/prefill.py:521          <- LLM prefill, long sequences
models/demos/gemma4/tt/attention/operations.py:398
models/demos/z_image_turbo/tt/dit/model_ttnn.py:291
models/demos/z_image_turbo/tt/vae/model_ttnn.py:439
models/tt_dit/models/vae/vae_sd35.py:322
models/demos/informer/tt/attention.py:445, :622
models/experimental/tt_symbiote/core/dispatchers/default_dispatcher.py:425
models/experimental/depth_anything_v2/tt/model_def.py:751
models/experimental/pi0/tt/ttnn_gemma.py:320
models/experimental/detr3d/ttnn/multihead_attention.py:70
```

The gemma4 prefill site is the one that matters: that is where sequence length runs into the thousands, which is where both gaps open. Still true with #50648 applied — it rewrites `gemma4/tt/attention/prefill.py` substantially but the SDPA call still passes only `compute_kernel_config`.

## Trade-offs a fix has to answer

**L1 headroom.** Bigger chunks mean bigger circular buffers, which share L1 with anything the model has resident. Ceiling, measured by pinning a sharded tensor until SDPA stops building:

| shape | max resident L1/core, default | derived | lost |
|---|---|---|---|
| seq 4096, dh 128 | 1312 KB | 928 KB | 384 KB |
| seq 8192, dh 128 | 1312 KB | 608 KB | 704 KB |
| seq 16384, dh 128 | 1312 KB | 576 KB | 736 KB |

That is the ceiling, not the common case. Free L1 at the actual call, via `ttnn.get_memory_view`, at gemma4's shapes:

| operands | free L1/bank | builds, default | derived |
|---|---|---|---|
| DRAM interleaved, 9 shape x seq combinations | 1427.1 KB | 9/9 | 9/9 |
| DRAM + an unrelated 256 KB/core pinned | 1171.1 KB | 9/9 | 9/9 |
| L1 interleaved | 235–1237 KB | 6/9 | 4/9 |

Every shipped call site is on the first row — `operations.py` says so outright, *"memory_config defaults to DRAM (true prefill: seq_len can be thousands of tokens and would not fit L1)"* — and sharded operands are rejected anyway. Worst CB footprint is 851 KB/core against 1427 KB free, so 576 KB/core of margin. The two cases that flip are L1-*interleaved* operands at 16q/8kv and 32q/16kv, where the operands alone take ~1192 KB/core; no default-path caller does that, and such a caller can pass a `program_config`.

**Sliding window.** The K loop is bounded in units of k-chunks, so a `k_chunk` wider than the window walks mostly-masked chunks. A fix that ignores `sliding_window_size` is **56% slower than the flat 32** at sequence 8192 with a window of 128 — the shape gemma4 prefill uses. A window needs two rules, not one: cap `k_chunk` at half the window, and — because a window bounds each Q chunk's K work by the window rather than by `Sk` — ask for a full grid of Q chunks rather than a quarter of one. Missing the second cost 38% at sequence 4096 with a window of 128. With both, the twelve window/length combinations run 1.04x to 3.91x against the flat 32, and none regresses.

**Causal accuracy vs speed.** A wider `q_chunk` reaches past its own diagonal, so its earliest rows fold through fully-masked k-chunks — no added mass, one more rounding each. At 8192 causal, `q_chunk` 128 is 1.4x faster than 64 and moves the worst deviation from 0.031 to 0.063 (against 0.277 for the flat 32). Both ends beat the default; which one belongs in a default is a judgement call.

Also: `chunk_start_idx` must divide both chunk sizes, which `validate()` enforces only for caller-supplied configs. Flexible chunked prefill and `cu_window_seqlens` describe their geometry with a device tensor the host cannot read, so those two paths cannot be derived from.

## Related

#21199 (closed) is the nearest existing report — degraded SDPA prefill PCC for Qwen 2.5 at sequence 4096, attributed there to the input distribution. It used an explicit `program_config`, so it is not this, but the symptom class is the same and the numbers above may be worth re-reading against it. #17793 and #25504 both ask about SDPA performance while passing explicit chunk sizes of 256/512 and 256/256 — neither touches the default.

## Why no test catches this

Every test in `tests/ttnn/unit_tests/operations/sdpa/` passes an explicit `program_config`. Nothing exercises the default path at all, so no golden encodes it and nothing fails. Worth closing regardless of what happens to the default.

## Please complete the following environment information

* OS: Ubuntu 22.04
* Version of software: `f6e166da2c`, sfpi 7.67.0
* Hardware: Blackhole P300, single chip, 11x10 grid

No Wormhole and no multi-device machine available; the L1 numbers are for a 1.5 MB L1.

**Comment by pavlejosipovic** [comment]
This is by design, defaults optimises for smallest L1 memory footprint it's a conservative default.
For perf you need to select optimal config that fits into you L1 memory budget.

**Comment by kinginu** [comment]
Understood on the L1 side, that makes sense as a default.

What I'm less clear on is the shipped models. gemma4's prefill calls SDPA without a `program_config`, so every attention layer takes the 32/32 path. Measured at its own shapes and layer mix, that comes to roughly 3.6x the SDPA time it needs: about 5.7 s against 1.6 s per forward pass at 16k context on the 31B config. There are 11 call sites under `models/` in the same position.

Is the expectation that each model picks its own config? If so, I'm happy to look at those call sites rather than the default.

On the accuracy side, for what it's worth: I checked the output against an fp32 reference and most of the difference turns out to be a per-row scale, so PCC barely moves (0.99932 against 0.99970 at sequence 16384). The row sums are still off, but I can't claim it changes model output, so I'd treat that as the weaker half of the report.

**Comment by pavlejosipovic** [comment]
That is the question for who ever is the code owner of that that model. My guess is they didn't get around to do perf rounds on prefill side.

**Comment by kinginu** [comment]
Thanks — I'll treat the default as working as intended and won't push further on it.
